src/core/agent/services/identifierService.ts

```diff
@@ -139,20 +161,57 @@ class IdentifierService extends AgentService {
     };
   }
 
+  async resolvePendingIdentifier() {
+    const pendingIdentifers =
+      await this.identifierStorage.getIdentifierPendingCreation();
+
+    for (const identifier of pendingIdentifers) {
+      await this.props.signifyClient
+        .identifiers()
+        .addEndRole(
+          identifier.id,
+          "agent",
+          this.props.signifyClient.agent!.pre
+        );
+    }
+  }
+
   @OnlineOnly
-  async createIdentifier(
+  async markIdentifierPendingCreate(
     metadata: Omit<IdentifierMetadataRecordProps, "id" | "createdAt">
-  ): Promise<CreateIdentifierResult> {
-    const startTime = Date.now();
-
+  ): Promise<void> {
     if (!identifierTypeThemes.includes(metadata.theme)) {
       throw new Error(`${IdentifierService.THEME_WAS_NOT_VALID}`);
     }
 
-    let name = `${metadata.theme}:${metadata.displayName}`;
-    if (metadata.groupMetadata) {
-      name = `${metadata.theme}:${metadata.groupMetadata.groupId}:${metadata.displayName}`;
-    }
+    const name = metadata.groupMetadata
+      ? `${metadata.theme}:${metadata.groupMetadata.groupId}:${metadata.displayName}`
+      : `${metadata.theme}:${metadata.displayName}`;
+
+    await this.basicStorage.createOrUpdateBasicRecord(
+      new BasicRecord({
+        id: MiscRecordId.IDENTIFIER_PENDING_CREATION,
+        content: {
+          name,
+        },
+      })
+    );
+
+    this.props.eventEmitter.emit<IdentifierCreatedEvent>({
+      type: EventTypes.IdentifierCreated,
+      payload: {
+        name,
+        metadata,
+      },
+    });
+  }
+
+  @OnlineOnly
+  async createIdentifier(
```

Comment:
Because of crashing and re-starting, it's possible that `createIdentifier` will get called for a certain `displayName` but only get half-way through before removing it from the basic record. So it'll try again.

We need to make this function more idempotent:
- If we try to create an identifier and it already exists by name, ignore the error and continue.
- If we try to save the pending operation and it already exists, ignore and continue, and skip emitting the event for the operation
- If we try to create the metadata record and it already exists, ignore and continue

This makes things look messy but otherwise we'll get in bad states.

Reply:
I’ve updated a few things to make the function more idempotent. Could you review it and provide details about any changes or ideas you’d like to implement if you’re not satisfied with my updates? I’ll make further updates today so we can complete this task.